pkg/liquidity-source/mantle/meth/pool_simulator.go

```diff
@@ -0,0 +1,148 @@
+	"encoding/json"
+	"errors"
+	"math/big"
+
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/entity"
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/liquidity-source/mantle/common"
+	poolpkg "github.com/KyberNetwork/kyberswap-dex-lib/pkg/source/pool"
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/util/bignumber"
+	"github.com/holiman/uint256"
+	"github.com/samber/lo"
+)
+
+type (
+	PoolSimulator struct {
+		poolpkg.Pool
+
+		isStakingPaused        bool
+		minimumStakeBound      *uint256.Int
+		maximumMETHSupply      *uint256.Int
+		totalControlled        *uint256.Int
+		exchangeAdjustmentRate uint16
+		mETHTotalSupply        *uint256.Int
+
+		gas Gas
+	}
+
+	Gas struct {
+		Stake int64
+	}
+)
+
+var (
+	ErrStakingPaused                 = errors.New("staking paused")
+	ErrorInvalidTokenIn              = errors.New("invalid tokenIn")
+	ErrorInvalidTokenOut             = errors.New("invalid tokenOut")
+	ErrMinimumStakeBoundNotSatisfied = errors.New("minimum stake bound not satisfied")
+	ErrMaximumMETHSupplyExceeded     = errors.New("maximum METH supply exceeded")
+	ErrStakeBelowMinimumMETHAmount   = errors.New("stake below minimum METH amount")
+)
+
+func NewPoolSimulator(entityPool entity.Pool) (*PoolSimulator, error) {
+	var extra PoolExtra
+	if err := json.Unmarshal([]byte(entityPool.Extra), &extra); err != nil {
+		return nil, err
+	}
+
+	return &PoolSimulator{
+		Pool: poolpkg.Pool{Info: poolpkg.PoolInfo{
+			Address:     entityPool.Address,
+			ReserveUsd:  entityPool.ReserveUsd,
+			Exchange:    entityPool.Exchange,
+			Type:        entityPool.Type,
+			Tokens:      lo.Map(entityPool.Tokens, func(item *entity.PoolToken, index int) string { return item.Address }),
+			Reserves:    lo.Map(entityPool.Reserves, func(item string, index int) *big.Int { return bignumber.NewBig(item) }),
+			BlockNumber: entityPool.BlockNumber,
+		}},
+		isStakingPaused:        extra.IsStakingPaused,
+		minimumStakeBound:      extra.MinimumStakeBound,
+		maximumMETHSupply:      extra.MaximumMETHSupply,
+		totalControlled:        extra.TotalControlled,
+		exchangeAdjustmentRate: extra.ExchangeAdjustmentRate,
+		mETHTotalSupply:        extra.METHTotalSupply,
+		gas:                    defaultGas,
+	}, nil
+}
+
+func (s *PoolSimulator) CalcAmountOut(params poolpkg.CalcAmountOutParams) (*poolpkg.CalcAmountOutResult, error) {
+	if s.isStakingPaused {
+		return nil, ErrStakingPaused
+	}
+
+	if params.TokenAmountIn.Token != WETH {
+		return nil, ErrorInvalidTokenIn
+	}
+
+	if params.TokenOut != METH {
+		return nil, ErrorInvalidTokenOut
+	}
+
+	amountIn := uint256.MustFromBig(params.TokenAmountIn.Amount)
+
+	if amountIn.Cmp(s.minimumStakeBound) < 0 {
+		return nil, ErrMinimumStakeBoundNotSatisfied
+	}
+
+	amountOut, err := s.ethToMETH(amountIn)
+	if err != nil {
+		return nil, err
+	}
+
+	if new(uint256.Int).Add(amountOut, s.mETHTotalSupply).Cmp(s.maximumMETHSupply) > 0 {
+		return nil, ErrMaximumMETHSupplyExceeded
+	}
+
+	if amountOut.Cmp(uint256.NewInt(0)) < 0 {
```

Comment:
this should compare with `minMETHAmount`, not `0`

Reply:
`minMETHAmount` is a fixed argument

https://github.com/KyberNetwork/ks-dex-aggregator-sc/blob/develop/src/contracts/executor-helpers/ExecutorHelper6.sol#L141

Reply:
ah I see. anyway can update this to 

```suggestion
	if amountOut.Sign() < 0 {
```

to save on 1 allocation